File: app/src/main/java/net/yanzm/fabtransformingsample/transition/CircularRevealTransition.java
package net.yanzm.fabtransformingsample.transition;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.TimeInterpolator;
import android.transition.TransitionValues;
import android.transition.Visibility;
import android.util.ArrayMap;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroup;

import java.util.ArrayList;

public class CircularRevealTransition extends Visibility {

    private final float startRadius;

    public CircularRevealTransition(float startRadius) {
        this.startRadius = startRadius;
    }

    public CircularRevealTransition(int mode, float startRadius) {
        this.startRadius = startRadius;
        setMode(mode);
    }

    @Override
    public Animator onAppear(ViewGroup sceneRoot, final View view, TransitionValues startValues,
                             TransitionValues endValues) {

        float endRadius = (float) (Math.hypot(view.getWidth(), view.getHeight()));

        int centerX = (int) view.getPivotX();
        int centerY = (int) view.getPivotY();

        view.setAlpha(0f);
        final Animator anim = ViewAnimationUtils.createCircularReveal(
                view, centerX, centerY, startRadius, endRadius);
        anim.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                super.onAnimationStart(animation);
                view.setAlpha(1f);
            }
        });
        return new NoPauseAnimator(anim);
    }

    @Override
    public Animator onDisappear(ViewGroup sceneRoot, final View view, TransitionValues startValues,
                                TransitionValues endValues) {
        float endRadius = (float) (Math.sqrt(view.getWidth() * view.getWidth() + view.getHeight() * view.getHeight()));

        int centerX = (int) view.getPivotX();
        int centerY = (int) view.getPivotY();

        final Animator anim = ViewAnimationUtils.createCircularReveal(
                view, centerX, centerY, endRadius, startRadius);
        return new NoPauseAnimator(anim);
    }

    private static class NoPauseAnimator extends Animator {

        private final Animator originalAnimator;
        private final ArrayMap<AnimatorListener, AnimatorListenerWrapper> listeners = new ArrayMap<>();

        NoPauseAnimator(Animator animator) {
            originalAnimator = animator;
        }

        @Override
        public void addListener(Animator.AnimatorListener listener) {
            final AnimatorListenerWrapper wrapper = new AnimatorListenerWrapper(this, listener);
            if (!listeners.containsKey(listener)) {
                listeners.put(listener, wrapper);
                originalAnimator.addListener(wrapper);
            }
        }

        @Override
        public void cancel() {
            originalAnimator.cancel();
        }

        @Override
        public void end() {
            originalAnimator.end();
        }

        @Override
        public long getDuration() {
            return originalAnimator.getDuration();
        }

        @Override
        public TimeInterpolator getInterpolator() {
            return originalAnimator.getInterpolator();
        }

        @Override
        public ArrayList<AnimatorListener> getListeners() {
            return new ArrayList<>(listeners.keySet());
        }

        @Override
        public long getStartDelay() {
            return originalAnimator.getStartDelay();
        }

        @Override
        public boolean isPaused() {
            return originalAnimator.isPaused();
        }

        @Override
        public boolean isRunning() {
            return originalAnimator.isRunning();
        }

        @Override
        public boolean isStarted() {
            return originalAnimator.isStarted();
        }

        @Override
        public void removeAllListeners() {
            super.removeAllListeners();
            listeners.clear();
            originalAnimator.removeAllListeners();
        }

        @Override
        public void removeListener(AnimatorListener listener) {
            AnimatorListener wrapper = listeners.get(listener);
            if (wrapper != null) {
                listeners.remove(listener);
                originalAnimator.removeListener(wrapper);
            }
        }

    /* We don't want to override pause or resume methods
     * because we don't want them to affect originalAnimator.
    public void pause();
    public void resume();
    public void addPauseListener(AnimatorPauseListener listener);
    public void removePauseListener(AnimatorPauseListener listener);
     */

        @Override
        public Animator setDuration(long durationMS) {
            originalAnimator.setDuration(durationMS);
            return this;
        }

        @Override
        public void setInterpolator(TimeInterpolator timeInterpolator) {
            originalAnimator.setInterpolator(timeInterpolator);
        }

        @Override
        public void setStartDelay(long delayMS) {
            originalAnimator.setStartDelay(delayMS);
        }

        @Override
        public void setTarget(Object target) {
            originalAnimator.setTarget(target);
        }

        @Override
        public void setupEndValues() {
            originalAnimator.setupEndValues();
        }

        @Override
        public void setupStartValues() {
            originalAnimator.setupStartValues();
        }

        @Override
        public void start() {
            originalAnimator.start();
        }

        private static class AnimatorListenerWrapper implements Animator.AnimatorListener {
            private final Animator animator;
            private final Animator.AnimatorListener originalListener;

            AnimatorListenerWrapper(Animator animator,
                                    Animator.AnimatorListener listener) {
                this.animator = animator;
                originalListener = listener;
            }

            @Override
            public void onAnimationStart(Animator animator) {
                originalListener.onAnimationStart(this.animator);
            }

            @Override
            public void onAnimationEnd(Animator animator) {
                originalListener.onAnimationEnd(this.animator);
            }

            @Override
            public void onAnimationCancel(Animator animator) {
                originalListener.onAnimationCancel(this.animator);
            }

            @Override
            public void onAnimationRepeat(Animator animator) {
                originalListener.onAnimationRepeat(this.animator);
            }
        }
    }
}

File: app/src/main/java/net/yanzm/fabtransformingsample/transition/CircularRevealTransition.kt
package net.yanzm.fabtransformingsample.transition

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.TimeInterpolator
import android.transition.TransitionValues
import android.transition.Visibility
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import java.util.ArrayList
import kotlin.math.hypot
import kotlin.math.sqrt

class CircularRevealTransition : Visibility {

    private val startRadius: Float

    constructor(startRadius: Float) {
        this.startRadius = startRadius
    }

    constructor(mode: Int, startRadius: Float) {
        this.startRadius = startRadius
        setMode(mode)
    }

    override fun onAppear(
        sceneRoot: ViewGroup,
        view: View,
        startValues: TransitionValues,
        endValues: TransitionValues
    ): Animator? {
        val w = view.width.toFloat()
        val h = view.height.toFloat()
        val endRadius = hypot(w, h)
        val centerX = view.pivotX.toInt()
        val centerY = view.pivotY.toInt()
        view.alpha = 0f
        val anim = ViewAnimationUtils.createCircularReveal(
            view, centerX, centerY, startRadius, endRadius
        ).apply {
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    super.onAnimationStart(animation)
                    view.alpha = 1f
                }
            })
        }
        return NoPauseAnimator(anim)
    }

    override fun onDisappear(
        sceneRoot: ViewGroup,
        view: View,
        startValues: TransitionValues,
        endValues: TransitionValues
    ): Animator? {
        val w = view.width.toFloat()
        val h = view.height.toFloat()
        val endRadius = sqrt(w * w + h * h)
        val centerX = view.pivotX.toInt()
        val centerY = view.pivotY.toInt()
        val anim = ViewAnimationUtils.createCircularReveal(
            view, centerX, centerY, endRadius, startRadius
        )
        return NoPauseAnimator(anim)
    }

    /**
     *  We don't want to override pause or resume methods
     * because we don't want them to affect originalAnimator.
     * public void pause();
     * public void resume();
     * public void addPauseListener(AnimatorPauseListener listener);
     * public void removePauseListener(AnimatorPauseListener listener);
     */
    private class NoPauseAnimator(private val originalAnimator: Animator) : Animator() {

        private val listeners = mutableMapOf<AnimatorListener, AnimatorListenerWrapper>()

        override fun addListener(listener: AnimatorListener) {
            val wrapper = AnimatorListenerWrapper(this, listener)
            if (!listeners.containsKey(listener)) {
                listeners[listener] = wrapper
                originalAnimator.addListener(wrapper)
            }
        }

        override fun cancel() {
            originalAnimator.cancel()
        }

        override fun end() {
            originalAnimator.end()
        }

        override fun getDuration(): Long {
            return originalAnimator.duration
        }

        override fun getInterpolator(): TimeInterpolator {
            return originalAnimator.interpolator
        }

        override fun getListeners(): ArrayList<AnimatorListener> {
            return ArrayList(listeners.keys)
        }

        override fun getStartDelay(): Long {
            return originalAnimator.startDelay
        }

        override fun isPaused(): Boolean {
            return originalAnimator.isPaused
        }

        override fun isRunning(): Boolean {
            return originalAnimator.isRunning
        }

        override fun isStarted(): Boolean {
            return originalAnimator.isStarted
        }

        override fun removeAllListeners() {
            super.removeAllListeners()
            listeners.clear()
            originalAnimator.removeAllListeners()
        }

        override fun removeListener(listener: AnimatorListener) {
            val wrapper: AnimatorListener? = listeners[listener]
            if (wrapper != null) {
                listeners.remove(listener)
                originalAnimator.removeListener(wrapper)
            }
        }

        override fun setDuration(durationMS: Long): Animator {
            originalAnimator.duration = durationMS
            return this
        }

        override fun setInterpolator(timeInterpolator: TimeInterpolator) {
            originalAnimator.interpolator = timeInterpolator
        }

        override fun setStartDelay(delayMS: Long) {
            originalAnimator.startDelay = delayMS
        }

        override fun setTarget(target: Any?) {
            originalAnimator.setTarget(target)
        }

        override fun setupEndValues() {
            originalAnimator.setupEndValues()
        }

        override fun setupStartValues() {
            originalAnimator.setupStartValues()
        }

        override fun start() {
            originalAnimator.start()
        }

        private class AnimatorListenerWrapper(
            private val animator: Animator,
            private val originalListener: AnimatorListener
        ) : AnimatorListener {
            override fun onAnimationStart(animator: Animator) {
                originalListener.onAnimationStart(this.animator)
            }

            override fun onAnimationEnd(animator: Animator) {
                originalListener.onAnimationEnd(this.animator)
            }

            override fun onAnimationCancel(animator: Animator) {
                originalListener.onAnimationCancel(this.animator)
            }

            override fun onAnimationRepeat(animator: Animator) {
                originalListener.onAnimationRepeat(this.animator)
            }
        }
    }
}
